### PR TITLE
Add design overview document

### PR DESCRIPTION
**Pull request by simonz-bq**
*Issue #, if available:*

https://bitquill.atlassian.net/browse/AN-1030

*Description of changes:*

Added a document outlining the overarching design and function of the gremlingo package.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: gremlin-go/design.md [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+-->
+
+# Gremlin-Go Design
+
+## Entities
+
+Gremlin-Go, named `gremlingo` as a package, consists of two main groups of entities at a high level: *Driver*-related and *Gremlin*-related entities. 
+
+### Driver
+
```
Comment (lyndonbauto) [inline gremlin-go/design.md:29]: Should we reference DriverRemoteConnection?

File: gremlin-go/design.md [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+Driver-related entities are used to handle the processing, parsing, and sending of Gremlin traversal queries. They are also responsible for deserializing responses from a Gremlin Server and the API for consuming the responses.
+
+The entities are as follows, from the highest level to the lowest:
+
+#### Client
+
+A `Client` represents the entry point to interaction with a Gremlin-supported server. It takes in basic configuration such as the hostname, port, as well as any additional configuration.
```
Comment (lyndonbauto) [inline gremlin-go/design.md]:
```suggestion
A `Client` represents the entry point to interaction with a Gremlin-supported server. It takes in a url, http headers, tls configuration, and additional configuration options.
```

File: gremlin-go/design.md [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+	loop Readloop
+		protocol->>protocol: Async population of ResultSet
+	end
+	connection-->>Client: ResultSet
+	
+```
+
+#### protocol
+
+The `protocol` entity handles invoking serialization and deserialization of data, as well as handling the lifecycle of raw data passed to and received from the `transporter` layer. Upon creation, an instance of `protocol` starts a `go routine` to asynchronously read and popular data into a  `ResultSet` that is owned by the parent `connection`.
```
Comment (lyndonbauto) [inline gremlin-go/design.md]:
```suggestion
The `protocol` entity handles invoking serialization and deserialization of data, as well as handling the lifecycle of raw data passed to and received from the `transporter` layer. Upon creation, an instance of `protocol` starts a `go routine` to asynchronously read and populate data into a  `ResultSet` that is owned by the parent `connection`.
```

File: gremlin-go/design.md (hunk 1 above)
Comment (lyndonbauto) [reply gremlin-go/design.md:29]: Oh I see, it is below. I am wondering if that should be more at the top because it is kind of our main entry point

Comment (simonz-bq) [reply gremlin-go/design.md:29]: I see your point, but I think given that this is a design doc and not a quickstart guide which highlights the `DriverRemoteConnection` I don't think it's necessary.

**Review by lyndonbauto — APPROVED** [review]

File: gremlin-go/design.md [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+	loop Readloop
+		protocol->>protocol: Async population of ResultSet
+	end
+	connection-->>Client: ResultSet
+	
+```
+
+#### protocol
+
+The `protocol` entity handles invoking serialization and deserialization of data, as well as handling the lifecycle of raw data passed to and received from the `transporter` layer. Upon creation, an instance of `protocol` starts a `go routine` to asynchronously read and populate data into a  `ResultSet` that is owned by the parent `connection`.
```
Comment (L0Lmaker) [inline gremlin-go/design.md]: nit:
```suggestion
The `protocol` entity handles invoking serialization and deserialization of data, as well as handling the lifecycle of raw data passed to and received from the `transporter` layer. Upon creation, an instance of `protocol` starts a `goroutine` to asynchronously read and populate data into a  `ResultSet` that is owned by the parent `connection`.
```

File: gremlin-go/design.md [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+##### Cardinalities
+
+* One `transporter`
+* One `serializer`
+
+##### Lifecycle and States
+
+* States
+  * `closed bool`
+* When `close()` is invoked, set the `closed` boolean to `true` which will terminate the `go routine` used for asynchronously reading, and invoke `close()` on the `transporter`.
```
Comment (L0Lmaker) [inline gremlin-go/design.md]: nit:
```suggestion
* When `close()` is invoked, set the `closed` boolean to `true` which will terminate the `goroutine` used for asynchronously reading, and invoke `close()` on the `transporter`.
```

File: gremlin-go/design.md [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+	ResultSet: Channel() chan
+	ResultSet: All() []Result
+	ResultSet: GetError() error
+```
+
+
+
+### Gremlin
+
+Gremlin-related entities are for the purpose of enabling the Gremlin query language to be used programmatically in Go. They responsible for integration with the Driver components, as well as translating the query language to a format that can be sent and consumed by a Gremlin-supported server. Most importantly, these entities are what allows the actual writing of Gremlin traversals in Go.
```
Comment (L0Lmaker) [inline gremlin-go/design.md]:
```suggestion
Gremlin-related entities are for the purpose of enabling the Gremlin query language to be used programmatically in Go. They are responsible for integration with the Driver components, as well as translating the query language to a format that can be sent and consumed by a Gremlin-supported server. Most importantly, these entities are what allow the actual writing of Gremlin traversals in Go.
```

File: gremlin-go/design.md [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+	class Result
+	Result: GetString() string
+	Result: GetInt() int
+	Result: GetByte() byte
+	Result: Get...() ...
+```
+
+#### ResultSet
+
+A `ResultSet` is the immediate output of executing a Gremlin traversal. It contains a set of the individual `Result` types. `ResultSet` is populated asynchronously by a go routine and thus also handles the providing of `Result` as they are available transparently to the user.
```
Comment (L0Lmaker) [inline gremlin-go/design.md]: nit:
```suggestion
A `ResultSet` is the immediate output of executing a Gremlin traversal. It contains a set of the individual `Result` types. `ResultSet` is populated asynchronously by a goroutine and thus also handles the providing of `Result` as they are available transparently to the user.
```

File: gremlin-go/design.md [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+  * `established`
+  * `closed`
+  * `closedDueToError`
+* When `close()` is invoked, set the state to `closed` and also invoke `close()` on the `protocol`.
+
+```mermaid
+classDiagram
+	class connection
+	connection: protocol *protocol
+	connection: results "map[string]ResultSet"
```
Comment (xiazcy) [inline gremlin-go/design.md:95]: Is there a reason `map[string]ResultSet` is in quotations?

File: gremlin-go/design.md (hunk 4 above)
Comment (xiazcy) [inline gremlin-go/design.md]:
```suggestion
The `protocol` entity handles invoking serialization and deserialization of data, as well as handling the lifecycle of raw data passed to and received from the `transporter` layer. Upon creation, an instance of `protocol` starts a `goroutine` to asynchronously read and populate data into a  `ResultSet` that is owned by the parent `connection`.
```

File: gremlin-go/design.md (hunk 5 above)
Comment (xiazcy) [inline gremlin-go/design.md]:
```suggestion
* When `close()` is invoked, set the `closed` boolean to `true` which will terminate the `goroutine` used for asynchronously reading, and invoke `close()` on the `transporter`.
```

File: gremlin-go/design.md [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+```mermaid
+classDiagram
+	class serializer
+	serializer: serializeMessage(request) []byte
+	serializer: deserializeMessage([]byte) response
+```
+
+#### transporter
+
+The `transporter` is an interface that describes the lowest-level methods that are required for sending and receiving requests, which implementing types are ones that are implementations of network protocols. The default implementation used for Gremlin-Go is Gorilla WebSocket, a Go implementation of the WebSocket protocol.
```
Comment (xiazcy) [inline gremlin-go/design.md]: I'd embed a link to Gorilla WebSocket repo or reference page here for `Gorilla WebSocket`

**Review by xiazcy — APPROVED** [review]
Just some nits and comments to be addressed.

File: gremlin-go/design.md (hunk 8 above)
Comment (simonz-bq) [reply gremlin-go/design.md:95]: Mermaid doesnt like go syntax, so when this is unquoted it tries to parse it and you can't escape it.

**Pull request merged**